### PR TITLE
[SQL Migration] Properly respect user's encryptConnection and trustServerCertificate settings

### PR DESCRIPTION
**Pull request by raymondtruong**
There has been an ongoing issue with the SQL Migration extension / DMS RP where various attempts to connect to the source instance (e.g. during IR validation, during the actual migration, and also during login / TDE migrations) are failing because the incorrect values of `encryptConnection` and `trustServerCertificate` are being used. This issue requires a multifaceted fix, including an RP hotfix deployment, as well as several changes to the behavior of the SQL Migration extension.
 
This PR:
- updates the appropriate RP request bodies (such as `validateIR` and `startMigration`) to explicitly provide values for  `encryptConnection` and `trustServerCertificate`, as provided by the user in ADS
  - previously, this was inconsistent. For example, when initially implemented, `encryptConnection = true` was hardcoded for `validateIR` (resulting in issues like https://github.com/microsoft/azuredatastudio/issues/21769), whereas for `startMigration`, it wasn't even being explicitly passed at all (and the RP was defaulting to `false`)
- fixes a bug where attempting to read the `encryptConnection` and `trustServerCertificate` values from the current connection (`ConnectionProfile` with the ADS API) would always result in `false`, even if the user provided otherwise. This is caused by an issue where ADS is returning these values as strings (`"true"`/`"false"`) instead of booleans (`true`/`false`) at times, whereas our extension was performing exclusively boolean comparisons. 
  - this is an underlying ADS bug (cc @cheenamalhotra?) where these values can be returned as either booleans or strings, depending on the connection. The current workaround is to both check for string `"true"` as well as boolean `true`. 
- remove references to stale connections. Currently, when the extension is initialized, we internally maintain a copy (actually, multiple copies) of the current connection (at least at the time of initialization), therefore, once connection settings are changed after the extension has been loaded, these changes will not take effect inside the extension. Additionally, it seems we were saving these stale references in a number of several different duplicate places as well. Storing the connection information locally is unnecessary and all such references have been replaced with calls to fetch the latest connection details whenever they're needed.
- vbumps the extension to 1.3.1
- refactors miscellaneous helper code to clean up

Screenshots:
- source ADS connection with encrypt connection `false`, trust server certificate `false`:
  - image
    - `validateIR` request body: 
    image
    - `startMigration` request body:
    image
- source ADS connection with encrypt connection `true`, trust server certificate `true`:
  - image
    - `validateIR` request body:
    image
    - `startMigration` request body:
    image

To-do:
- verify what values of `encryptConnection` and `trustServerCertificate` should be used when connecting to **target** servers, SQL DB/MI/VM, including for login and TDE migration - currently I have not modified any of the existing behavior

### CODE REVIEW

File: extensions/sql-migration/src/api/azure.ts [hunk 1]
```diff
@@ -654,17 +655,18 @@ export async function validateIrSqlDatabaseMigrationSettings(
...
 			trustServerCertificate: trustServerCertificate,
-			// encryptConnection: true,
 		},
 		targetSqlConnection: {
 			testConnectivity: testTargetConnectivity,
 			dataSource: targetDatabaseServer.properties.fullyQualifiedDomainName,
 			userName: migration._targetUserName,
 			password: migration._targetPassword,
+			authentication: MigrationSourceAuthenticationType.Sql,
+			// when connecting to a target Azure SQL DB, use true/false
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/azure.ts:667]: IR validation for SQL DB scenario - verify that true/true is correct for the target

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 2]
```diff
@@ -153,6 +179,7 @@ function getSqlDbConnectionProfile(
 			password: password,
 			connectionTimeout: 60,
 			columnEncryptionSetting: 'Enabled',
+			// when connecting to a target Azure SQL DB, use true/false
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/sqlUtils.ts:182]: collecting target database table info for SQL DB scenario - verify that true/true is correct for the target

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 3]
```diff
@@ -203,8 +231,35 @@ export function getConnectionProfile(
...
+	azureResourceId: string,
+	username: string,
+	password: string): Promise<string> {
+
+	const connectionProfile = getTargetConnectionProfile(
+		serverName,
+		azureResourceId,
+		username,
+		password,
+		// for login migration, when connecting to a target Azure SQL, use true/true
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/sqlUtils.ts:249]: for login migration scenario, verify that true/true is correct for the target (regardless of target platform)

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 4]
```diff
@@ -269,11 +324,14 @@ export async function collectTargetDatabaseInfo(
...
 	password: string): Promise<TargetDatabaseInfo[]> {
 
-	const connectionProfile = getConnectionProfile(
+	const connectionProfile = getTargetConnectionProfile(
 		targetServer.properties.fullyQualifiedDomainName,
 		targetServer.id,
 		userName,
-		password);
+		password,
+		// when connecting to a target Azure SQL DB, use true/false
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/sqlUtils.ts:332]: collecting target database table info for SQL DB scenario - verify that true/false is correct for the target

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 5]
```diff
@@ -394,11 +452,13 @@ export async function collectTargetLogins(
 	password: string,
 	includeWindowsAuth: boolean = true): Promise<string[]> {
 
-	const connectionProfile = getConnectionProfile(
+	const connectionProfile = getTargetConnectionProfile(
 		serverName,
 		azureResourceId,
 		userName,
 		password,
+		// for login migration, when connecting to a target Azure SQL, use true/true
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/sqlUtils.ts:460]: for login migration scenario, verify that true/true is correct for the target (regardless of target platform)

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 6]
```diff
@@ -1103,7 +1084,7 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 		reportUpdate: (dbName: string, succeeded: boolean, message: string) => Promise<void>): Promise<OperationResult<TdeMigrationDbResult[]>> {
 
 		const tdeEnabledDatabases = this.tdeMigrationConfig.getTdeEnabledDatabases();
-		const connectionString = await azdata.connection.getConnectionString(this.sourceConnectionId, true);
+		const connectionString = await getSourceConnectionString();
```
Comment (raymondtruong) [inline extensions/sql-migration/src/models/stateMachine.ts:1087]: for TDE migration, verify that no functionality has changed

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 7]
```diff
@@ -1166,8 +1144,8 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 					authentication: this._authenticationType,
 					userName: this._sqlServerUsername,
 					password: this._sqlServerPassword,
-					// to-do: use correct value of encryptConnection and trustServerCertificate
-					trustServerCertificate: currentConnection?.options.trustServerCertificate ?? false
+					encryptConnection: currentConnection?.options.encrypt === true || currentConnection?.options.encrypt === 'true',
```
Comment (raymondtruong) [inline extensions/sql-migration/src/models/stateMachine.ts:1147]: for `startMigration`, verify that both `encryptConnection` and `trustServerCertificate` are explicitly passed

File: extensions/sql-migration/src/api/azure.ts [hunk 8]
```diff
@@ -728,7 +731,7 @@ export async function validateIrDatabaseMigrationSettings(
 			dataSource: sourceServerName,
 			userName: migration._sqlServerUsername,
 			password: migration._sqlServerPassword,
-			// to-do: use correct value of encryptConnection and trustServerCertificate
+			encryptConnection: encryptConnection,
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/azure.ts:734]: for `validateIR` for MI/VM, verify that both `encryptConnection` and `trustServerCertificate` are explicitly passed

File: extensions/sql-migration/src/api/azure.ts [hunk 9]
```diff
@@ -654,17 +655,18 @@ export async function validateIrSqlDatabaseMigrationSettings(
 			userName: migration._sqlServerUsername,
 			password: migration._sqlServerPassword,
 			authentication: migration._authenticationType,
+			encryptConnection: encryptConnection,
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/azure.ts:658]: for `validateIR` for DB, verify that both `encryptConnection` and `trustServerCertificate` are explicitly passed

File: extensions/sql-migration/src/api/azure.ts (hunk 1 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/azure.ts:667]: cc @siyangMicrosoft

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 2 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:182]: cc @siyangMicrosoft

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 3 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:249]: cc @AkshayMata

File: extensions/sql-migration/src/api/azure.ts (hunk 1 above)
Comment (neetusingh-ms) [reply extensions/sql-migration/src/api/azure.ts:667]: https://learn.microsoft.com/en-us/azure/azure-sql/database/connect-query-content-reference-guide?view=azuresql#tls-considerations-for-database-connectivity

recommendation is 'true' always.

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 4 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:332]: cc @siyangMicrosoft

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 5 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:460]: cc @AkshayMata

File: extensions/sql-migration/src/models/stateMachine.ts (hunk 6 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/models/stateMachine.ts:1087]: cc @junierch

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 10]
```diff
@@ -127,6 +127,32 @@ export interface LoginTableInfo {
 	status: string;
 }
 
+export async function getSourceConnectionProfile(): Promise<azdata.connection.ConnectionProfile> {
```
Comment (raymondtruong) [inline extensions/sql-migration/src/api/sqlUtils.ts:130]: Majority of the changes in this PR revolve around refactoring these commonly used helpers, and replacing references to possibly-stale connection info with these calls to always fetch the latest connection info

File: extensions/sql-migration/src/api/azure.ts (hunk 1 above)
Comment (AkshayMata) [reply extensions/sql-migration/src/api/azure.ts:667]: yeah, Encrypt to "true" and TrustServerCertification to "false" should always work for MI and Azure SQL DB

File: extensions/sql-migration/src/api/azure.ts (hunk 8 above)
Comment (AkshayMata) [reply extensions/sql-migration/src/api/azure.ts:734]: This is for the source tho right? So we should honor what the user has specified?

File: extensions/sql-migration/src/api/azure.ts (hunk 1 above)
Comment (brian-harris) [reply extensions/sql-migration/src/api/azure.ts:667]: We should only /always connect to Azure targets using encrypt on and trust off.

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 3 above)
Comment (AkshayMata) [reply extensions/sql-migration/src/api/sqlUtils.ts:249]: Thanks for checking in! I love the way you clearly outlines all the scenarios and risk factors!

To answer your question, the doc above applies to "Azure SQL Database" and "Azure SQL Managed Instance." So for MI the default true/false will work. However, that doesn't work for SQL VM, which is why I had to hard code it in the first place. 

The long term solution should be:
- For SQL MI, default to true/false. 
- For SQL VM, we give customer UX dropdowns to select how they want to connect.

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 11]
```diff
@@ -203,8 +231,35 @@ export function getConnectionProfile(
 	};
 }
 
-export async function collectSourceDatabaseTableInfo(sourceConnectionId: string, sourceDatabase: string): Promise<TableInfo[]> {
-	const ownerUri = await azdata.connection.getUriForConnection(sourceConnectionId);
+export async function getSourceConnectionString(): Promise<string> {
+	return await azdata.connection.getConnectionString((await getSourceConnectionProfile()).connectionId, true);
+}
+
+export async function getTargetConnectionString(
```
Comment (AkshayMata) [inline extensions/sql-migration/src/api/sqlUtils.ts:238]: Should we make this function more generic? Right now it is hard-coded for login migration settings. There is a risk that someone might use this in the future based on the function name without realize the effects on encrypt/trustservercertificate

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 5 above)
Comment (AkshayMata) [reply extensions/sql-migration/src/api/sqlUtils.ts:460]: See comment above. To recap here the long term solution is:
- For SQL MI, default to true/false. 
- For SQL VM, we give customer UX dropdowns to select how they want to connect.

File: extensions/sql-migration/src/dashboard/dashboardTab.ts [hunk 12]
```diff
@@ -764,11 +765,10 @@ export class DashboardTab extends TabBase<DashboardTab> {
 			})
 			.component();
 
-		const connectionProfile = await azdata.connection.getCurrentConnection();
 		this.disposables.push(
 			this.serviceContextChangedEvent.event(
 				async (e) => {
-					if (e.connectionId === connectionProfile.connectionId) {
+					if (e.connectionId === await getSourceConnectionId()) {
```
Comment (brian-harris) [inline extensions/sql-migration/src/dashboard/dashboardTab.ts:771]: due to the implementation of this event, the code gets executed in all open instances of the extension... i.e. one per server connection.
 This logic was added to ensure that the Connection ID sending the event was the same one receiving it...  To test open up multiple instances/connections to two or more servers, then set different migration service contexts...  then ensure it doesn't change for the other open sql-migration dashboards related to/opened on different server connections

File: extensions/sql-migration/src/api/azure.ts (hunk 1 above)
Comment (AkshayMata) [reply extensions/sql-migration/src/api/azure.ts:667]: @brian-harris that doesn't work for SQL VM as the target platform

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 13]
```diff
@@ -127,6 +127,32 @@ export interface LoginTableInfo {
...
+export async function getSourceConnectionProfile(): Promise<azdata.connection.ConnectionProfile> {
+	return await azdata.connection.getCurrentConnection();
+}
+
+export async function getSourceConnectionId(): Promise<string> {
+	return (await getSourceConnectionProfile()).connectionId;
+}
+
+export async function getSourceConnectionServerInfo(): Promise<azdata.ServerInfo> {
+	return await azdata.connection.getServerInfo((await getSourceConnectionProfile()).connectionId);
```
Comment (brian-harris) [inline extensions/sql-migration/src/api/sqlUtils.ts:139]: can these methos be updated to replace (await getSourceConnectionProfile()).connectionId with  getSourceConnectionId etc

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 2 above)
Comment (brian-harris) [reply extensions/sql-migration/src/api/sqlUtils.ts:182]: we should only connect to azure targets with encrypt and no trust

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 13 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:139]: thanks, good catch

File: extensions/sql-migration/src/api/sqlUtils.ts [hunk 14]
```diff
@@ -203,8 +231,35 @@ export function getConnectionProfile(
...
+	password: string): Promise<string> {
+
+	const connectionProfile = getTargetConnectionProfile(
+		serverName,
+		azureResourceId,
+		username,
+		password,
+		// for login migration, when connecting to a target Azure SQL, use true/true
+		true /* encryptConnection */,
+		true /* trustServerCertificate */);
```
Comment (brian-harris) [inline extensions/sql-migration/src/api/sqlUtils.ts:251]: I don't think trust=true is correct, we should not trust cert connections to azure?

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 2 above)
Comment (brian-harris) [reply extensions/sql-migration/src/api/sqlUtils.ts:182]: same for other connections below...

File: extensions/sql-migration/src/api/azure.ts (hunk 8 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/azure.ts:734]: yes, on the source side in this example we do honor what the user passes in

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 11 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:238]: good call, will take the two bools as parameters

File: extensions/sql-migration/src/api/azure.ts (hunk 1 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/azure.ts:667]: resolving this thread in particular, does seem true/false for DB is correct

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 2 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:182]: looks like true/false was correct for DB and MI, resolving this thread, but will address true/false not working for VM elsewhere

File: extensions/sql-migration/src/api/sqlUtils.ts (hunk 4 above)
Comment (raymondtruong) [reply extensions/sql-migration/src/api/sqlUtils.ts:332]: true/false is correct

File: extensions/sql-migration/src/dialog/validationResults/validateIrDialog.ts [hunk 15]
```diff
@@ -363,11 +364,10 @@ export class ValidateIrDialog {
...
 
 	private async _validateDatabaseMigration(): Promise<void> {
-		const sqlConnections = await azdata.connection.getConnections();
-		const currentConnection = sqlConnections.find(
-			value => value.connectionId === this._model.sourceConnectionId);
+		const currentConnection = await getSourceConnectionProfile();
 		const sourceServerName = currentConnection?.serverName!;
-		const trustServerCertificate = currentConnection?.options?.trustServerCertificate === true;
+		const encryptConnection = currentConnection?.options?.encrypt === true || currentConnection?.options?.encrypt === 'true';
+		const trustServerCertificate = currentConnection?.options?.trustServerCertificate === true || currentConnection?.options?.trustServerCertificate === 'true';
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/validationResults/validateIrDialog.ts:370]: nit, maybe create helper methods to reading each of these?

Comment (raymondtruong) [reply extensions/sql-migration/src/dialog/validationResults/validateIrDialog.ts:370]: good call

**Review by brian-harris — APPROVED** [review]

**Pull request merged**